transport/src/test/java/io/camunda/zeebe/transport/stream/impl/StreamIntegrationTest.java

```diff
@@ -237,18 +196,194 @@ private AtomixCluster createClusterNode(final Node localNode, final Collection<N
+  @Nested
+  final class LifecycleTest {
+    @Test
+    void shouldAddRemoteStream() {
+      // given
+      final var streamType = BufferUtil.wrapString("foo");
+      final var properties = new TestSerializableData();
+
+      // when
+      final var streamId =
+          clientStreamer
+              .add(streamType, properties, p -> CompletableFuture.completedFuture(null))
+              .join();
+
+      // then
+      awaitStreamAdded(streamType, streamId);
+    }
+
+    @Test
+    void shouldRemoveRemoteStream() {
+      // given
+      final var streamType = BufferUtil.wrapString("foo");
+      final var properties = new TestSerializableData();
+      final var streamId =
+          clientStreamer
+              .add(streamType, properties, p -> CompletableFuture.completedFuture(null))
+              .join();
+
+      // must wait until the stream is connected everywhere before removal, as otherwise there is a
+      // race condition
+      awaitStreamAdded(streamType, streamId);
+
+      // when
+      clientStreamer.remove(streamId).join();
+
+      // then
+      awaitStreamOnServer(streamType, server1, stream -> assertThat(stream).isEmpty());
+      awaitStreamOnServer(streamType, server2, stream -> assertThat(stream).isEmpty());
+      awaitStreamOnClient(streamId, stream -> assertThat(stream).isEmpty());
+      awaitStreamOnClient(streamId, stream -> assertThat(stream).isEmpty());
+    }
+
+    @Test
+    void shouldDisconnectFromDeadServer() {
+      // given
+      final var streamType = BufferUtil.wrapString("foo");
+      final var properties = new TestSerializableData();
+      final var streamId =
+          clientStreamer
+              .add(streamType, properties, p -> CompletableFuture.completedFuture(null))
+              .join();
+      awaitStreamAdded(streamType, streamId);
+
+      // when
+      client.streamService.onServerRemoved(server1.memberId());
+
+      // then
+      awaitStreamOnClient(
+          streamId,
+          stream -> {
+            assertThat(stream).isPresent();
+            //noinspection OptionalGetWithoutIsPresent
+            assertThat(stream.get().isConnected(server1.memberId())).isFalse();
+            assertThat(stream.get().isConnected(server2.memberId())).isTrue();
+          });
+    }
+
+    @Test
+    void shouldRegisterStreamOnReconnectedServer() {
+      // given
+      final var streamType = BufferUtil.wrapString("foo");
+      final var properties = new TestSerializableData();
+      client.streamService.onServerRemoved(server1.memberId());
+      final var streamId =
+          clientStreamer
+              .add(streamType, properties, p -> CompletableFuture.completedFuture(null))
+              .join();
+      awaitStreamOnServer(streamType, server2, stream -> assertThat(stream).isPresent());
+      awaitStreamOnClient(
+          streamId,
+          stream -> {
+            assertThat(stream).isPresent();
+            //noinspection OptionalGetWithoutIsPresent
```

Comment:
:X: I think you want to make sure that server1 is removed before re-adding it.
```suggestion
            //noinspection OptionalGetWithoutIsPresent
            assertThat(stream.get().isConnected(server1.memberId())).isFalse();
```

Reply:
Good catch, that was 100% copy paste :smile: